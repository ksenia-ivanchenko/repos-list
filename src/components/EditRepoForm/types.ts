import { TRepo } from "../../store/slices/repo/types";

export interface TEditRepoFormProps {
    handleEdit: (values) => void;
    repo: TRepo;
}
