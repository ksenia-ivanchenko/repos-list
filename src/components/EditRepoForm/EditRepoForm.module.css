.form {
    padding: 20px 10px 0;
    display: flex;
    flex-direction: column;
}

.submit {
    align-self: flex-end;
}
