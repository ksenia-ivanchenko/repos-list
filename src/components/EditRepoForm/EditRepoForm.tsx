import { Button, Form, Input } from "antd";
import { TEditRepoFormProps } from "./types";
import { FC } from "react";
import styles from "./EditRepoForm.module.css";

export const EditRepoForm: FC<TEditRepoFormProps> = ({ repo, handleEdit }) => (
    <Form
        data-testid="edit-form"
        name="edit"
        onFinish={handleEdit}
        className={styles.form}
        initialValues={{ repo: repo.name, owner: repo.owner.login }}
    >
        <Form.Item label="Название репозитория" name="repo">
            <Input />
        </Form.Item>

        <Form.Item label="Имя владельца" name="owner">
            <Input />
        </Form.Item>

        <Form.Item label={null} className={styles.submit}>
            <Button type="primary" htmlType="submit">
                Сохранить
            </Button>
        </Form.Item>
    </Form>
);
