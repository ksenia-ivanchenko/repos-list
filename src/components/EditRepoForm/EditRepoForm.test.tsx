import { EditRepoForm } from "./EditRepoForm";
import {
    fireEvent,
    render,
    screen,
    act,
    cleanup,
} from "@testing-library/react";
import * as reduxHooks from "react-redux";
import { mockGetRepo } from "../../tests/mock/mockGetRepos";

jest.mock("react-redux");

const mockedUseDispatch = jest.spyOn(reduxHooks, "useDispatch");

describe("EditRepoForm", () => {
    afterEach(() => cleanup());

    it("should create EditRepoForm", () => {
        const component = render(
            <EditRepoForm repo={mockGetRepo.items[0]} handleEdit={jest.fn()} />
        );

        expect(component).toMatchSnapshot();
    });

    it("should correctly render EditRepoForm", () => {
        render(
            <EditRepoForm repo={mockGetRepo.items[0]} handleEdit={jest.fn()} />
        );

        expect(screen.getByTestId("edit-form")).toBeInTheDocument();
        expect(document.getElementById("edit_repo")).toBeInTheDocument();
        expect(document.getElementById("edit_owner")).toBeInTheDocument();
        expect(
            screen.getByRole("button", { name: /сохранить/i })
        ).toBeInTheDocument();
    });

    it("should dispatch when submit and call handleEdit", async () => {
        const handleEditMock = jest.fn();
        mockedUseDispatch.mockReturnValue(jest.fn());

        render(
            <EditRepoForm
                repo={mockGetRepo.items[0]}
                handleEdit={handleEditMock}
            />
        );

        fireEvent.change(screen.getByLabelText(/название репозитория/i), {
            target: { value: "новое название" },
        });
        fireEvent.change(screen.getByLabelText(/имя владельца/i), {
            target: { value: "новый владелец" },
        });

        await act(async () => {
            fireEvent.submit(screen.getByTestId("edit-form"));
        });

        expect(handleEditMock).toHaveBeenCalledTimes(1);
        expect(handleEditMock).toHaveBeenCalledWith({
            repo: "новое название",
            owner: "новый владелец",
        });
    });

    it("should close modal after successful submit", async () => {
        const handleEditMock = jest.fn().mockResolvedValueOnce(true);

        render(
            <EditRepoForm
                repo={mockGetRepo.items[0]}
                handleEdit={handleEditMock}
            />
        );

        fireEvent.change(screen.getByLabelText(/название репозитория/i), {
            target: { value: "новое название" },
        });

        fireEvent.submit(screen.getByTestId("edit-form"));

        await act(async () => {
            await handleEditMock();
        });

        expect(screen.queryByTestId("edit-modal")).toBeNull();
    });
});
