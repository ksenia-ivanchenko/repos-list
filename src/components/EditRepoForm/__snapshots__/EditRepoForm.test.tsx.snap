// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRepoForm should create EditRepoForm 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-horizontal css-dev-only-do-not-override-nqoqt9 form"
        data-testid="edit-form"
        id="edit"
      >
        <div
          class="ant-form-item css-dev-only-do-not-override-nqoqt9"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-nqoqt9"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-nqoqt9"
            >
              <label
                class=""
                for="edit_repo"
                title="Название репозитория"
              >
                Название репозитория
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-nqoqt9"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input css-dev-only-do-not-override-nqoqt9 ant-input-outlined"
                    id="edit_repo"
                    type="text"
                    value="bootstrap"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item css-dev-only-do-not-override-nqoqt9"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-nqoqt9"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-nqoqt9"
            >
              <label
                class=""
                for="edit_owner"
                title="Имя владельца"
              >
                Имя владельца
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-nqoqt9"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input css-dev-only-do-not-override-nqoqt9 ant-input-outlined"
                    id="edit_owner"
                    type="text"
                    value="twbs"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item submit css-dev-only-do-not-override-nqoqt9"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-nqoqt9"
          >
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-nqoqt9"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-nqoqt9 ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                    type="submit"
                  >
                    <span>
                      Сохранить
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-nqoqt9 form"
      data-testid="edit-form"
      id="edit"
    >
      <div
        class="ant-form-item css-dev-only-do-not-override-nqoqt9"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-nqoqt9"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-nqoqt9"
          >
            <label
              class=""
              for="edit_repo"
              title="Название репозитория"
            >
              Название репозитория
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-nqoqt9"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input css-dev-only-do-not-override-nqoqt9 ant-input-outlined"
                  id="edit_repo"
                  type="text"
                  value="bootstrap"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-nqoqt9"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-nqoqt9"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-nqoqt9"
          >
            <label
              class=""
              for="edit_owner"
              title="Имя владельца"
            >
              Имя владельца
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-nqoqt9"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input css-dev-only-do-not-override-nqoqt9 ant-input-outlined"
                  id="edit_owner"
                  type="text"
                  value="twbs"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item submit css-dev-only-do-not-override-nqoqt9"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-nqoqt9"
        >
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-nqoqt9"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-nqoqt9 ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                  type="submit"
                >
                  <span>
                    Сохранить
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
